Make get_range_addresses async and hold effective_replication_map_ptr around it

bhalevy opened the pull request:
This series converts the synchronous `effective_replication_map::get_range_addresses` to async
by calling the replication strategy async entry point with the same name, as its callers are already async
or can be made so easily.

To allow it to yield and work on a coherent view of the token_metadata / topology / replication_map,
let the callers of this patch hold a effective_replication_map per keyspace and pass it down
to the (now asynchronous) functions that use it (making affected storage_service methods static where possible
if they no longer depend on the storage_service instance).

Also, the repeated calls to everywhere_replication_strategy::calculate_natural_endpoints
are optimized in this series by introducing a virtual abstract_replication_strategy::has_static_natural_endpoints predicate
that is true for local_strategy and everywhere_replication_strategy, and is false otherwise.
With it, functions repeatedly calling calculate_natural_endpoints in a loop, for every token, will call it only once since it will return the same result every time anyhow.

Refs #11005 

Doesn't fix the issue as the large allocation still remains until we make change dht::token_range_vector chunked (chunked_vector cannot be used as is at the moment since we require the ability to push also to the front when unwrapping)

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -256,18 +256,9 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
-        }
-    }
-    return ret;
+    return _rs->get_range_addresses(*_tmptr);
```
1. Can it be precomputed? Probably wasteful since it's rarely needed
2. Can it be optimized by merging precomputed results of get_natural_endpoint()? Probably we do have get_natural_endpoints(t) cached.
3. Can it be improved for everywhere_replication_strategy by making its effective_replication_map single-range (ignoring vnodes)? Probably

bhalevy replied:
> 1. Can it be precomputed? Probably wasteful since it's rarely needed

Agreed.

> 2. Can it be optimized by merging precomputed results of get_natural_endpoint()? Probably we do have get_natural_endpoints(t) cached.

Hmm, yes.  Good catch.
We can use the replication_map directly, as `get_natural_endpoints` simply searches the replication map
or calls the replication strategy specific api.  The replication map is already pre-calculated with the natural endpoint so we should be able to safely use it.

> 3. Can it be improved for everywhere_replication_strategy by making its effective_replication_map single-range (ignoring vnodes)? Probably

Agreed. but that's a far bigger change.

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -169,6 +169,12 @@ effective_replication_map::get_ranges(inet_address ep) const {
     });
 }
 
```
changelog type: `s/only once it/only once if it/`

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -178,9 +184,11 @@ abstract_replication_strategy::get_ranges(inet_address ep, token_metadata_ptr tm
     auto prev_tok = sorted_tokens.back();
     for (auto tok : sorted_tokens) {
-        for (inet_address a : co_await calculate_natural_endpoints(tok, tm)) {
+        co_await maybe_calculate_natural_endpoints(eps, tok, tm);
```
I think that for everywhere and local, ignoring vnodes will yield a much bigger benefit (especially everywhere which is quadratic with vnodes).

bhalevy replied:
I tend to agree but there are edge cases where we currently return the local endpoint while token_metadata is empty and after it is populated we return all nodes.  To do that we'll probably need to override several abstract_replication_strategy methods in everywhere r.s. and local strategy so they'll have their simplified implementation.

Maybe I'll get back to it when working on compaction groups.

avikivity replied:
Can you elaborate? Why is token_metadata empty?

bhalevy replied:
IIRC when bootstrapping a node token_metadata is initially empty until gossip decides about the bootstrap tokens

bhalevy replied:
Another point of concern is that the token_metadata is currently shared to all replication strategies and there is a single (versioned) source of truth for it, and it is sampled by the effective_,replication_map.  For having a single token for local/everywhere strategies we'll need to decouple their token_metadata from the rest.  It's probably possible, but will require delicate surgery.

bhalevy commented:
In v2 (7aa186fbc0):
- added "effective_replication_map: get_range_addresses: use the precalculated replication_map"
- fixed typo in "abstract_replication_strategy: add maybe_calculate_natural_endpoints" commit log message

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1265/

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -260,9 +260,8 @@ std::unordered_map<dht::token_range, inet_address_vector_replica_set>
-    for (auto& t : tm.sorted_tokens()) {
+    for (const auto& [t, eps] : _replication_map) {
         dht::token_range_vector ranges = tm.get_primary_ranges_for(t);
-        auto eps = get_natural_endpoints(t);
```
Ah, that's the real fix.

avikivity commented on service/storage_service.cc:
```diff
@@ -2767,7 +2767,7 @@ future<> storage_service::removenode_add_ranges(lw_shared_ptr<dht::range_streame
             }
         }
-        std::unordered_multimap<inet_address, dht::token_range> source_ranges = get_new_source_ranges(keyspace_name, my_new_ranges);
+        std::unordered_multimap<inet_address, dht::token_range> source_ranges = co_await get_new_source_ranges(keyspace_name, my_new_ranges);
```
Do we know that the token metadata won't change while we calculate this?

bhalevy replied:
Addressed in v3

avikivity commented on service/storage_service.cc:
```diff
@@ -3011,12 +3011,12 @@ future<> storage_service::shutdown_protocol_servers() {
     auto& ks = _db.local().find_keyspace(keyspace_name);
     auto erm = ks.get_effective_replication_map();
-    std::unordered_map<dht::token_range, inet_address_vector_replica_set> range_addresses = erm->get_range_addresses();
+    std::unordered_map<dht::token_range, inet_address_vector_replica_set> range_addresses = co_await erm->get_range_addresses();
```
When we return, erm will be outdated.

bhalevy replied:
Done in v3

bhalevy commented on locator/abstract_replication_strategy.cc:
```diff
@@ -294,8 +294,16 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
+            if (ep == pending_address) {
+                dht::token_range_vector r = temp.get_primary_ranges_for(t);
+                rslogger.debug("get_pending_address_ranges: token={} primary_range={} endpoint={}", t, r, pending_address);
+                std::copy(r.begin(), r.end(), std::back_inserter(ret));
```
I presume it's reserving the required space in advance.

avikivity replied:
No, it doesn't, std::copy() is too decoupled from back_inserter_iterator.

std::vector::insert(iterator, iterator, iterator) should work though.

bhalevy replied:
Done in v3

bhalevy commented on locator/abstract_replication_strategy.cc:
```diff
@@ -294,8 +294,16 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
+                rslogger.debug("get_pending_address_ranges: token={} primary_range={} endpoint={}", t, r, pending_address);
+                std::copy(r.begin(), r.end(), std::back_inserter(ret));
+                break;
+            }
```
we append the result of `get_primary_ranges_for(t)` to `ret` only if `pending_address` is contained in `calculate_natural_endpoints(t, temp);`

bhalevy replied:
Re-looking at it, I think we better move the implementation into the effective replication map since there there is no need to calculate_natural_endpoints for every token. It's already pre-calculated in the replication map.

avikivity replied:
Ah. Can we write it as `if (eps.contains(pending_address)`? It's much clearer.

bhalevy replied:
In fact I started working today on a change to caculate_natural_endpoints to return a endpoint_set which is a utils::sequenced_set, i.e. a set/vector hybrid which can be used both ways.  It is built internally in network_topology_replication_strategy anyhow today.

bhalevy commented:
In v3 (1d52266140):
- Cleanup and generalization of utils:sequenced_set
  - "utils: sequenced_set: delete mutable get_vector method"
  - "utils: sanitize sequenced_set"
  - "utils: sequenced_set: templatize VectorType"
  - "utils: sequenced_set: expose set and `contains` method"
- "abstract_replication_strategy: calculate_natural_endpoints: return endpoint_set"
  - improves performance of users consuming `calculate_natural_endpoints`
- "abstract_replication_strategy: get_pending_address_ranges: prevent extra vector copies"
  - uses vector::insert
- effective_replication_map lifetime cleanup:
  - "effective_replication_map: add get_replication_strategy"
  - "database: add get_non_system_keyspaces_erms"
  - "storage_service: get_ranges_for_endpoint: get erm as param"
  - "storage_service: get_changed_ranges_for_leaving: get erm as param"
  - "storage_service: get_new_source_ranges: get erm as param"
  - "range_streamer: add_ranges and friends: get erm as param"
  - "range_streamer: add_ranges and friends: get erm as param"
    - made safe WRT using e_r_m with the above

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1331/

bhalevy commented:
Editing the series... Will post v4 soon

bhalevy commented:
In v4 (64959faa06):
- storage_service: coroutinize update_pending_ranges
- database: add get_non_local_strategy_keyspaces
  - and use it tree-wide
- database: add get_non_local_strategy_keyspaces_erms
  - renamed
  - repair: use get_non_local_strategy_keyspaces_erms
  - and used in more places

bhalevy commented:
@asias please review too, especially the repair-related patches.

avikivity commented:
Does it overlap with #10850 (which I missed before)?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1347/

bhalevy commented:
> Does it overlap with #10850 (which I missed before)?

@avikivity, yes it does, we'll have to rebase one on top of the other to resolve merge conflicts.

bhalevy commented:
In 34a4970149 - rebased on top of #10850
No functional changes.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1555/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1555/

Apparently this was caused by an unrelated issue.
See https://github.com/scylladb/scylla-tools-java/pull/303

bhalevy commented:
1749070283: rebased to get the CI fix

bhalevy commented:
In v5 (7183caa5f3)
- rebased + retested
- fixed doxygen comment in storage_service.hh

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1636/

avikivity commented on locator/abstract_replication_strategy.hh:
```diff
@@ -44,6 +45,8 @@ using replication_strategy_config_options = std::map<sstring, sstring>;
 
 using replication_map = std::unordered_map<token, inet_address_vector_replica_set>;
 
+using endpoint_set = utils::basic_sequenced_set<inet_address, inet_address_vector_replica_set>;
```
I'm not thrilled about sequenced_set, it was a type used to emulate some Java type, not something with deep meaning. But okay.

avikivity commented on api/storage_service.cc:
```diff
@@ -548,7 +548,7 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
     ss::describe_any_ring.set(r, [&ctx, &ss](std::unique_ptr<request> req) {
         // Find an arbitrary non-system keyspace.
-        auto keyspaces = ctx.db.local().get_non_system_keyspaces();
+        auto keyspaces = ctx.db.local().get_non_local_strategy_keyspaces();
```
Absolutely, s/accidentally correct thing/meaningful API/

avikivity commented on replica/database.hh:
```diff
@@ -1496,6 +1496,7 @@ public:
     std::vector<sstring> get_user_keyspaces() const;
     std::vector<sstring> get_all_keyspaces() const;
     std::vector<sstring> get_non_local_strategy_keyspaces() const;
+    std::unordered_map<sstring, locator::effective_replication_map_ptr> get_non_local_strategy_keyspaces_erms() const;
```
_erms() is not friendly to newcomers, especially since the return type likely won't be seen at the call site. Please expand.

avikivity commented on repair/repair.cc:
```diff
@@ -1342,17 +1342,16 @@ future<> repair_service::bootstrap_with_repair(locator::token_metadata_ptr tmptr
         seastar::sharded<replica::database>& db = get_db();
-        auto keyspaces = db.local().get_non_local_strategy_keyspaces();
+        auto ks_erms = db.local().get_non_local_strategy_keyspaces_erms();
         auto myip = utils::fb_utilities::get_broadcast_address();
```
Very good.

Idea: add bool effective_replication_map::outdated() const that becomes true if replication changed, so we can abort repair. Soon we'll start waiting for all users of an old topology to complete before moving to a new topology, and this will allow us to move quicker.

bhalevy replied:
sounds good. we should be able to mark the e_r_m as outdated in O(1) when it's replaced in the registry.

avikivity commented on service/storage_service.cc:
```diff
@@ -3347,8 +3352,8 @@ storage_service::get_splits(const sstring& ks_name, const sstring& cf_name, rang
-    return _db.local().find_keyspace(name).get_effective_replication_map()->get_ranges(ep);
+storage_service::get_ranges_for_endpoint(const locator::effective_replication_map_ptr& erm, const gms::inet_address& ep) const {
+    return erm->get_ranges(ep);
 }
```
Really the function can just go away, no?

Or you could make it static to emphasize it's pure and doesn't depend on any state outside the input.

bhalevy replied:
I'd like to reduce the churn, so making it static sounds good.

avikivity commented on service/storage_service.cc:
```diff
@@ -2656,10 +2656,7 @@ int32_t storage_service::get_exception_count() {
-    // FIXME: get erm as param
-    auto& ks = _db.local().find_keyspace(keyspace_name);
-    auto erm = ks.get_effective_replication_map();
+future<std::unordered_multimap<dht::token_range, inet_address>> storage_service::get_changed_ranges_for_leaving(locator::effective_replication_map_ptr erm, inet_address endpoint) {
```
Did this also become a pure function? If so it should be static (or free)

bhalevy replied:
I think that with `get_ranges_for_endpoint` becoming static this one could potentially become static too.

avikivity commented on service/storage_service.cc:
```diff
@@ -3004,10 +3004,8 @@ future<> storage_service::shutdown_protocol_servers() {
-    auto erm = ks.get_effective_replication_map();
     std::unordered_map<dht::token_range, inet_address_vector_replica_set> range_addresses = erm->get_range_addresses();
     std::unordered_multimap<inet_address, dht::token_range> source_ranges;
 
```
Ok, I see it calls _gossiper.is_alive(), so these mix run-time information with the inputs. So can't change here.

avikivity commented:
Looks good, I left minor comments. 

IMO the nesting of users of topology within holders of effective_replication_map_ptr that the series does is more important than fixing the stalls.

bhalevy commented:
> IMO the nesting of users of topology within holders of effective_replication_map_ptr that the series does is more important than fixing the stalls.

I agree. It became more evident after your first review.

bhalevy commented:
In v6 (52188c861ac6b3286372d4565210f298e942092f):
- rebased + retested
- database: add get_non_local_strategy_keyspaces_and_effective_replication_maps
  - renamed from get_non_local_strategy_keyspaces_erms
- storage_service: get_ranges_for_endpoint: get erm as param and make static
  - also, get rid of now needless try/catch as the keyspace is no longer looked up by name now that `get_ranges_for_endpoint` is passed an effective_replication_map rather than the keyspace name
  - storage_service: effective_ownership: fixup indentation
- storage_service: get_changed_ranges_for_leaving: get erm as param and make static
- storage_service: make get_broadcast_address static
  - while at it
  - though it's insufficient to allow get_new_source_ranges to become static
  - as the latter still depends on the _gossiper member (we could pass it to it but that seems awkward)

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -169,6 +169,12 @@ effective_replication_map::get_ranges(inet_address ep) const {
+future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(std::optional<endpoint_set>& eps, const token& search_token, const token_metadata& tm) const {
+    if (!eps || !has_static_natural_endpoints()) {
+        eps = co_await calculate_natural_endpoints(search_token, tm);
+    }
```
For Everywhere, don't the endpoints change when nodes are added or removed?

bhalevy replied:
Yes, but the topology never changes in the context of the eps, it holds an immutable token_metadata (that encapsulates the topology)

bhalevy replied:
But I see you point that this happens outside the scope of this function

bhalevy replied:
Agreed.  I'm about to revise it using a safe approach that detects if the token_metadata changed using its version. Although it shouldn't none of the current uses in this series, the method is exposed as public and external users may not respect that convention.

```diff
diff --git a/locator/abstract_replication_strategy.cc b/locator/abstract_replication_strategy.cc
index 4ffee1b10e..f49b69985c 100644
--- a/locator/abstract_replication_strategy.cc
+++ b/locator/abstract_replication_strategy.cc
@@ -169,9 +169,12 @@ effective_replication_map::get_ranges(inet_address ep) const {
     });
 }
 
-future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(std::optional<endpoint_set>& eps, const token& search_token, const token_metadata& tm) const {
-    if (!eps || !has_static_natural_endpoints()) {
-        eps = co_await calculate_natural_endpoints(search_token, tm);
+future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(cached_endpoint_set& cached_eps, const token& search_token, const token_metadata& tm) const {
+    auto cur_ring_version = tm.get_ring_version();
+    assert(cur_ring_version >= 0);
+    if (!has_static_natural_endpoints() || cached_eps.ring_version != cur_ring_version) {
+        cached_eps.eps = co_await calculate_natural_endpoints(search_token, tm);
+        cached_eps.ring_version = cur_ring_version;
     }
 }
 
@@ -184,7 +187,7 @@ abstract_replication_strategy::get_ranges(inet_address ep, token_metadata_ptr tm
     if (sorted_tokens.empty()) {
         on_internal_error(rslogger, "Token metadata is empty");
     }
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
     auto prev_tok = sorted_tokens.back();
     for (auto tok : sorted_tokens) {
         co_await maybe_calculate_natural_endpoints(eps, tok, tm);
@@ -223,7 +226,7 @@ effective_replication_map::get_primary_ranges_within_dc(inet_address ep) const {
 future<std::unordered_multimap<inet_address, dht::token_range>>
 abstract_replication_strategy::get_address_ranges(const token_metadata& tm) const {
     std::unordered_multimap<inet_address, dht::token_range> ret;
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
 
     for (auto& t : tm.sorted_tokens()) {
         dht::token_range_vector r = tm.get_primary_ranges_for(t);
@@ -244,7 +247,7 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
     if (!tm.is_member(endpoint)) {
         co_return ret;
     }
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
     bool is_everywhere_topology = get_type() == replication_strategy_type::everywhere_topology;
     for (auto& t : tm.sorted_tokens()) {
         // This is a fast path for everywhere_topology to avoid calculating
@@ -265,7 +268,7 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
                 ret.emplace(endpoint, rng);
             }
         } else {
-            rslogger.debug("token={} natural_endpoints={}: endpoint={} not found", t, eps, endpoint);
+            rslogger.debug("token={} natural_endpoints={}: endpoint={} not found", t, *eps, endpoint);
         }
     }
     co_return ret;
@@ -288,7 +291,7 @@ effective_replication_map::get_range_addresses() const {
 future<std::unordered_map<dht::token_range, inet_address_vector_replica_set>>
 abstract_replication_strategy::get_range_addresses(const token_metadata& tm) const {
     std::unordered_map<dht::token_range, inet_address_vector_replica_set> ret;
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
 
     for (auto& t : tm.sorted_tokens()) {
         dht::token_range_vector ranges = tm.get_primary_ranges_for(t);
@@ -311,7 +314,7 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
     token_metadata temp;
     temp = co_await tmptr->clone_only_token_map();
     co_await temp.update_normal_tokens(pending_tokens, pending_address);
-    std::optional<endpoint_set> eps;
+    cached_endpoint_set eps;
 
     for (const auto& t : temp.sorted_tokens()) {
         co_await maybe_calculate_natural_endpoints(eps, t, temp);
@@ -327,7 +330,7 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
 
 future<mutable_effective_replication_map_ptr> calculate_effective_replication_map(abstract_replication_strategy::ptr_type rs, token_metadata_ptr tmptr) {
     replication_map replication_map;
-    std::optional<endpoint_set> eps;
+    abstract_replication_strategy::cached_endpoint_set eps;
 
     for (const auto &t : tmptr->sorted_tokens()) {
         co_await rs->maybe_calculate_natural_endpoints(eps, t, *tmptr);
diff --git a/locator/abstract_replication_strategy.hh b/locator/abstract_replication_strategy.hh
index f37ec38698..982915f067 100644
--- a/locator/abstract_replication_strategy.hh
+++ b/locator/abstract_replication_strategy.hh
@@ -82,7 +82,29 @@ class abstract_replication_strategy {
 
     virtual bool has_static_natural_endpoints() const noexcept { return false; }
 
-    future<> maybe_calculate_natural_endpoints(std::optional<endpoint_set>& eps, const token& search_token, const token_metadata& tm) const;
+    // Avoids redundant calculations
+    struct cached_endpoint_set {
+        endpoint_set eps;
+        long ring_version = -1;
+
+        const endpoint_set& operator*() const noexcept {
+            return eps;
+        }
+
+        endpoint_set& operator*() noexcept {
+            return eps;
+        }
+
+        const endpoint_set* operator->() const noexcept {
+            return &eps;
+        }
+
+        endpoint_set* operator->() noexcept {
+            return &eps;
+        }
+    };
+
+    future<> maybe_calculate_natural_endpoints(cached_endpoint_set& eps, const token& search_token, const token_metadata& tm) const;
 
     // The returned vector has size O(number of normal token owners), which is O(number of nodes in the cluster).
     // Note: it is not guaranteed that the function will actually yield. If the complexity of a particular implementation
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1657/

bhalevy commented:
In v6 (e6ddf8ea5a):
- abstract_replication_strategy: add maybe_calculate_natural_endpoints
  - define and use `cached_endpoint_set` to detect if the caller's token_metadata changed to recalculate the natural_endpoint, even if the replication strategy has_static_natural_endpoints()

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1671/

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -169,6 +169,15 @@ effective_replication_map::get_ranges(inet_address ep) const {
 
+future<> abstract_replication_strategy::maybe_calculate_natural_endpoints(cached_endpoint_set& cached_eps, const token& search_token, const token_metadata& tm) const {
+    auto cur_ring_version = tm.get_ring_version();
+    assert(cur_ring_version >= 0);
```
I'm confused. token_metadata_ptr already is the embodiment of a ring topology mixed into a replication strategy. Here you are replicating that mechanism with another cache on top of it.

I don't understand what's the goal of maybe_calculate_natural_endpoints. It looks like it's optimizing the easy case and rare cases, local or everywhere, and leaving the common and hard cases unoptimized.

bhalevy replied:
The goal is indeed to optimize the local and everywhere cases while keeping the other cases unoptimized.
`abstract_replication_strategy::calculate_natural_endpoints` is meant to be called on arbitrary token_metadata.
We use it typically on token_metadata that was cloned and possibly manipulated so it doesn't belong to any effective_replication_map - hence the natural endpoints for it aren't already precalculated.

In this patch `maybe_calculate_natural_endpoints` is used only internally by `abstract_replication_strategy`.
But I plan to use it in repair based node ops.  Maybe they should build an effective_replication_map (that is aware of this optimization) and use that instead of calling the abstract_replication_strategy calculate_natural_endpoints.

bhalevy replied:
@avikivity I don't want this last patch hold up the whole series.
Are you ok with defining the method private to abstract_replication_strategy and use it only internally there for now?

avikivity replied:
Sorry, I lost the thread. How about you drop this last patch and resurrect it in a separate series? I think it's not crucial for what the series does.

bhalevy replied:
done in v7

bhalevy commented:
In v7 (91ab8ee1c)
- Dropped `abstract_replication_strategy: add maybe_calculate_natural_endpoints`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1720/
